cmd/runner/main.go

```diff
@@ -53,16 +70,294 @@ func run(args []string) error {
+	// Make sure the destination exists
+	if err := os.MkdirAll(filepath.Dir(destPath), 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	destF, err := os.Create(destPath)
+	if err != nil {
+		return fmt.Errorf("failed to create dest file: %w", err)
+	}
+	defer destF.Close() // Best-effort in case something fails
 
-	h := taskrunner.New(&local.Executor{}, logger)
+	br, err := h.blob.ReadBlob(ctx, srcURI)
+	if err != nil {
+		return fmt.Errorf("failed to read raw portfolio: %w", err)
+	}
+	defer br.Close() // Best-effort in case something fails
+
+	if _, err := io.Copy(destF, br); err != nil {
+		return fmt.Errorf("failed to load raw portfolio: %w", err)
+	}
+
+	if err := br.Close(); err != nil {
+		return fmt.Errorf("failed to close blob reader: %w", err)
+	}
+
+	if err := destF.Close(); err != nil {
+		return fmt.Errorf("failed to close dest file: %w", err)
+	}
+
+	return nil
+}
+
+func (h *handler) processPortfolio(ctx context.Context, req *task.ProcessPortfolioRequest) error {
+	// Load the portfolio from blob storage, place it in /mnt/raw_portfolios, where
+	// the `process_portfolios.R` script expects it to be.
+	for _, assetID := range req.AssetIDs {
+		srcURI := blob.Join(h.blob.Scheme(), h.sourcePortfolioContainer, assetID)
+		// TODO: Probably set the CSV extension in the signed upload URL instead.
+		destPath := filepath.Join("/", "mnt", "raw_portfolios", assetID+".csv")
+
+		if err := h.downloadBlob(ctx, srcURI, destPath); err != nil {
+			return fmt.Errorf("failed to download raw portfolio blob: %w", err)
+		}
+	}
+
+	processedDir := filepath.Join("/", "mnt", "processed_portfolios")
+	if err := os.MkdirAll(processedDir, 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	var stdout, stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "/usr/local/bin/Rscript", "/app/process_portfolios.R")
+	cmd.Stdout = io.MultiWriter(os.Stdout, &stdout)
+	cmd.Stderr = io.MultiWriter(os.Stderr, &stderr)
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run process_portfolios script: %w", err)
+	}
+
+	sc := bufio.NewScanner(&stderr)
 
-	portfolioID := pacta.PortfolioID(os.Getenv("PORTFOLIO_ID"))
-	logger.Info("running PACTA task", zap.String("portfolio_id", string(portfolioID)))
-	if err := h.Execute(ctx, &task.StartRunRequest{
-		PortfolioID: portfolioID,
-	}); err != nil {
-		return fmt.Errorf("failed to run task: %w", err)
+	// TODO: Load from the output database file (or similar, like reading the processed_portfolios dir) instead of parsing stderr
+	var paths []string
+	for sc.Scan() {
+		line := sc.Text()
+		idx := strings.Index(line, "writing to file: " /* 17 chars */)
+		if idx == -1 {
+			continue
+		}
+		paths = append(paths, strings.TrimSpace(line[idx+17:]))
+	}
+
+	// NOTE: This code could benefit from some concurrency, but I'm opting not to prematurely optimize.
```

Comment:
Same thought here as above, maybe less important, since the cardinality is O(report outputs) not O(portfolio inputs).

Reply:
Same comment as above, my knowledge of what the actual data steps look like is fuzzy, and I don't know what the discrete units of compute are